### DISCUSSION TITLE
Client-side TanStack Router with Jotai

### DISCUSSION THREAD

**Discussion by fnn** [pmndrs/jotai#1847] [Q&A]
Hey,

I'm trying to figure out a good pattern to use Jotai with client-side only TanStack Router: https://github.com/TanStack/router/blob/v0.0.1-beta.83/docs/guide/data-loading.md

I kind of want to replicate the remix loader pattern. My simplest idea was to use `getDefaultStore()` inside the loader and use the `store` API to set an Atom. Then inside my component I'm able use `useAtomValue` to get my loader data. But I'm not sure if that is even an intended use-case without any side-effects.

I appreciate feedback an other ideas.

**Comment by fnn** [comment]
In my use-case I want to share my `react-query` queries from my loader with the components, so that I have just one place to edit queries. Maybe something like this:

```js
const loaderAtom = atom(null);

new Route({
  getParentRoute: () => rootRoute,
  path: "foobar",
  onLoad: async ({ context, search }) => {
    const query = createQuery({
      filter1: "value1",
      filter2: "value2",
    });

    const store = getDefaultStore();
    store.set(loaderAtom, {
      queries: {
        foobar: query,
      },
    });

    await context.queryClient.ensureQueryData(query);
  },
  component: function Foobar() {
    const data = useAtomValue(loaderAtom);
    const query = useQuery(data.queries.foobar);

    return <div>{JSON.stringify(data)}</div>;
  },
});
```

But it doesn't feel right.. Maybe use `tanstack-router`'s context for DI. Also getting the typings done feels weird.
There is probably a more jotaish way to do this kind of stuff.

EDIT: Also doesn't seem to work correctly, the Foobar component will get the changes from `useAtomValue(loaderAtom)`, but any child component will get `null` from that.

**Reply by dai-shi** [reply]
Are you sure you are using the default store in all child components? (I mean you don't use Jotai's `Provider`.)

**Reply by fnn** [reply]
Ah yes, you are right. Thanks!

**Comment by dai-shi** [comment]
> But I'm not sure if that is even an intended use-case without any side-effects.

What I can say is, it's why store api is exposed to cover such use cases..
We don't know it's going to be Jotai best practice yet. Still to explore.

**Reply by fnn** [reply]
That is good to hear. So I'm not completely off. I will complete my router migration and then, with a clear mind, continue experimenting with TanStack Router and Jotai.

**Comment by dai-shi** [comment]
I remember @t6adev did some experiments with store api and some router.

**Reply by t6adev** [reply]
Yup, I did with React Router's loader.
https://twitter.com/t6adev/status/1590524570150539264

**Reply by fnn** [reply]
Thanks, I am going to look into this and continue to experiment.
